timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+
+    """
+    host_bps, host_masses = openmm_deserializer.deserialize_system(host_config.omm_system, cutoff=1.2)
+    host_conf = minimizer.minimize_host_4d(
+        [st.mol_a, st.mol_b],
+        host_config.omm_system,
+        host_config.conf,
+        st.ff,
+        host_config.box,
+    )
+
+    all_U_fns = []
+    all_frames = []
+    all_boxes = []
+
+    for lamb in lambda_schedule:
+        hgs = st.combine_with_host(convert_bps_into_system(host_bps), lamb=lamb)
+        all_U_fns.append(hgs.get_U_fns())
+        bound_impls = [p.bound_impl(np.float32) for p in hgs.get_U_fns()]
+        # minimize water box around the ligand by 4D-decoupling
+        mol_a_conf = get_romol_conf(st.mol_a)
+        mol_b_conf = get_romol_conf(st.mol_b)
+        ligand_conf = st.combine_confs(mol_a_conf, mol_b_conf)
+        combined_conf = np.concatenate([host_conf, ligand_conf])
+        x0 = combined_conf
+        v0 = np.zeros_like(x0)
+        box0 = host_config.box
+        group_idxs = get_group_indices(get_bond_list(hgs.bond))
+        seed = np.random.randint(0, 1000)
+        combined_masses = np.concatenate([host_masses, st.combine_masses()])
+        frames, boxes = sample(bound_impls, combined_masses, x0, v0, box0, lamb, group_idxs, seed, n_frames=n_frames)
+        all_frames.append(frames)
+        all_boxes.append(boxes)
+
+    return all_frames, all_boxes, all_U_fns
+
+
+def plot_BAR(dG, dG_err, fwd_delta_u, rev_delta_u, title, plot_idx):
+    """
+    Generate a subplot showing overlap for a particular pair of delta_us.
+    """
+    plt.subplot(*plot_idx)
+    plt.title(f"{title}, dG: {dG:.2f} +- {dG_err:.2f} kTs")
+    plt.hist(fwd_delta_u, alpha=0.5, label="fwd", density=True, bins=20)
+    plt.hist(-rev_delta_u, alpha=0.5, label="-rev", density=True, bins=20)
+    plt.xlabel("work (kTs)")
+    plt.legend()
+
+
+SimulationResult = namedtuple("SimulationResult", ["all_dGs", "all_errs", "plot_png", "frames", "boxes", "U_fns"])
+
+
+def estimate_free_energy_given_samples(all_frames, all_boxes, all_U_fns, lambda_schedule, prefix):
+    """
+    Estimate free energies given pre-generated samples.Let L be the number of lambda windows,
+    T be the number of samples, N be the number of atoms, P be the number of components
+    in the energy function.
+
+    Parameters
+    ----------
+    all_frames: np.array of shape (L, T, N, 3)
+        List of coordinates from all windows
+
+    all_boxes: np.array of shape (L, T, 3, 3)
+        List of boxes from all windows
+
+    all_U_fns: list of energy functions of shape (L, P)
+        List of energy functions from all windows
+
+    lambda_schedule: np.array of shape (L,)
+        Lambda schedule
+
+    prefix: str
+        A prefix to append to overlap figures
+
+    Return
+    ------
+    3-tuple of dGs, dG_errs, overlap_plot
+        All components have have shape (L-1,)
+
+    """
+    # assume pair-BAR format
+    # setup U_fns to operate in batch mode
+    kT = BOLTZ * 300.0
```

Comment:
`timemachine.constants.DEFAULT_KT`

Reply:
I need to fix this to be self-consistent with the temperature of the thermostat.